File: flattenTreeToLL.cpp
// Time Complexity : O(n) will visit all nodes once
// Space Complexity : O(h) height of tree
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

// Your code here along with comments explaining your approach
// 1. Recursively call flatten for left subtree. Store right subtree address in temp
// 2. Make right of root equal to flattened left subtree and make left subtree as nullptr
// 3. Go to end of added flattened tree and place the right subtree
// 4. Flatten the right subtree  

class Solution {
public:
    void flatten(TreeNode* root) {
        // base
        if(root==nullptr || (root->left==nullptr && root->right==nullptr))
            return;
        // logic
        flatten(root->left);
        TreeNode* temp = root->right;
        root->right = root->left;
        root->left = nullptr;
        while(root->right != nullptr)
            root = root->right;
        root->right = temp;
        flatten(root->right);
    }
};

File: FlattenTreeToLL.py
# TC: O(n)
# SC: O(n)
# DFS: flatten subtrees at lower levels- rewire to make right skewed subtree, then proceed to parent branch- repeat until whole tree is a right skewed tree.
class Solution:
    def flatten(self, root: TreeNode) -> None:
        if not root: return root
        return self.helper(root)
    
    def helper(self, root):
        # base
        if not root: return None
        if root.left == None and root.right == None:
            return root
        
        # logics
        leftLeaf = self.helper(root.left)
        rightLeaf = self.helper(root.right)
        
        if leftLeaf:
            leftLeaf.right = root.right
            root.right = root.left
            root.left = None
        
        return rightLeaf if rightLeaf else leftLeaf